### PR TITLE
Cleanup edit groups after coalescing

### PR DESCRIPTION
**Pull request by dsnet**
Even with an optimal diffing algoritm, coalescing adjacent edit groups
may cause the corresponding pair of strings for an edit group to have
leading or trailing spans of equal elements.

While this is technically a correct representation of a diff,
it is a suboptimal outcome. As such, scan through all unequal groups and
move leading/trailing equal spans to the preceding/succeeding equal group.

Before this change:
```patch
  strings.Join({
    "org-4747474747474747,bucket-4242424242424242:m,tag1=a,tag2=aa",
-   ",#=_value _value=2 ",
+   " _value=2 ",
    `11 org-4747474747474747,bucket-4242424242424242:m,tag1=a,tag2=bb`,
-   ",#=_value _value=2 2",
+   " _value=2 2",
    `1  org-4747474747474747,bucket-4242424242424242:m,tag1=b,tag2=cc`,
-   ",#=_value",
    ` _value=1 21 org-4747474747474747,bucket-4242424242424242:m,tag1`,
    "=a,tag2",
-   "=dd,#=_value",
+   "=dd",
    ` _value=3 31 org-4747474747474747,bucket-4242424242424242:m,tag1`,
-   "=c,#=_value",
+   "=c",
    ` _value=4 41 `,
  }, "")
```

After this change:
```patch
  strings.Join({
    "org-4747474747474747,bucket-4242424242424242:m,tag1=a,tag2=aa",
-   ",#=_value",
    ` _value=2 11 org-4747474747474747,bucket-4242424242424242:m,tag1`,
    "=a,tag2=bb",
-   ",#=_value",
    ` _value=2 21 org-4747474747474747,bucket-4242424242424242:m,tag1`,
    "=b,tag2=cc",
-   ",#=_value",
    ` _value=1 21 org-4747474747474747,bucket-4242424242424242:m,tag1`,
    "=a,tag2=dd",
-   ",#=_value",
    ` _value=3 31 org-4747474747474747,bucket-4242424242424242:m,tag1`,
    "=c",
-   ",#=_value",
    ` _value=4 41 `,
  }, "")
```

### CODE REVIEW

**Comment by dsnet** [comment]
\cc @rogpeppe, this PR will fix the sub-optimal printing of the diffs.

**Review by rogpeppe — COMMENTED** [review]
LGTM - this looks like it'll be a huge improvement, thanks!
Note: I haven't looked at the detail of the implementation properly I'm afraid, but I'm very happy to see this change.

File: cmp/report_slices.go [hunk 1]
```diff
@@ -463,3 +507,91 @@ func coalesceInterveningIdentical(groups []diffStats, windowSize int) []diffStat
...
+				ix += numLeadingIdentical
+				iy += numLeadingIdentical
+			}
+			if numTrailingIdentical > 0 {
+				// Remove trailing identical span from this group and
+				// insert it into the succeeding group.
+				if i+1 < len(groups) {
+					groups[i+1].NumIdentical += numTrailingIdentical
+				} else {
+					// No succeeding group exist, so append a new group,
```
[the remark this answers is not in the document]
Comment (dsnet) [reply cmp/report_slices.go:579]: Done.

File: cmp/report_slices.go [hunk 2]
```diff
@@ -463,3 +507,91 @@ func coalesceInterveningIdentical(groups []diffStats, windowSize int) []diffStat
...
+			numTrailingIdentical++
+		}
+		if numIdentical := numLeadingIdentical + numTrailingIdentical; numIdentical > 0 {
+			if numLeadingIdentical > 0 {
+				// Remove leading identical span from this group and
+				// insert it into the preceding group.
+				if i-1 >= 0 {
+					groups[i-1].NumIdentical += numLeadingIdentical
+				} else {
+					// No preceding group exist, so prepend a new group,
```
[the remark this answers is not in the document]
Comment (dsnet) [reply cmp/report_slices.go]: Done.

**Review by neild — APPROVED** [review]

**Review by AnomalRoil — COMMENTED** [review]
I know this is already merged, but since I'm reviewing our dependencies before updating, I'm just letting you know my thoughts ;)

Thanks for the comments, it helped a lot with the review process.

**Comment by dsnet** [comment]
@AnomalRoil, I commend you for [reviewing changes to your dependencies](https://research.swtch.com/deps#code_quality)! This is something I wish people would do more often. Thank you.

**Comment by AnomalRoil** [comment]
Yup, and that's especially true when dealing with crypto & passwords, since these are quite sensitive. 
So I try to do it for [Gopass](https://github.com/gopasspw/gopass/).

**Pull request merged**